src/Lean/Compiler/JoinPoints.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2022 Henrik Böving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Lean.Compiler.CompilerM
+
+namespace Lean.Compiler
+
+namespace JoinPointChecker
+
+def jpArity (jp : LocalDecl) : Nat :=
+  go jp.value
+where
+  go : Expr → Nat
+  | .lam _ _ body _ => 1 + go body
+  | _ => 0
+
+/--
+Throws an exception if `e` contains a join point.
+-/
+partial def containsNoJp (e : Expr) : CompilerM Unit := do
+  match e.consumeMData with
+  | .proj _ _ struct => containsNoJp struct
+  | .lam .. =>
+    let (_, b) ← visitLambda e
+    containsNoJp b
+  | .letE .. =>
+    let body ← visitLet e (fun e => do containsNoJp e; pure e)
+    containsNoJp body
+  | .app fn arg =>
+    containsNoJp fn
+    containsNoJp arg
+  | .fvar fvarId =>
+    let some decl ← findDecl? fvarId | unreachable!
+    if decl.isJp then
+      throwError s!"Join point {decl.userName} in forbidden position"
+  | .sort .. | .forallE .. | .const .. | .lit .. => return ()
+  | .bvar .. | .mvar .. | .mdata .. => unreachable! 
+
+/--
+Check whether all join points in `e` are in a valid position that is:
+1. Fully applied
+2. In tail position inside of `e`
+-/
+partial def checkJoinPoints (e : Expr) :  CompilerM Unit := do
+  let (_, body) ← visitLambda e
+  go body
+where
+  goLetValue (value : Expr) : CompilerM Unit := do
+    let value := value.consumeMData
+    match value with
+    | .lam .. =>
+      let (_, body) ← visitLambda value
```

Comment:
We can use `withNewScope do` to make sure the local context is restored after we visit the lambda.
```lean
withNewScope do
  let (_, body) ← visitLambda value
  go body
```
For this checker, this change has very little (or no) impact, but it makes sure our code looks uniform.